File: mixtour/MIXViewController.m
//
//  MIXViewController.m
//  mixtour
//
//  Created by Carsten Wenderdel on 2013-06-17.
//  Copyright (c) 2013 Carsten Wenderdel. All rights reserved.
//

#import "MIXViewController.h"

#import "MIXGameView.h"
#import "MIXModelBoard.h"


@implementation MIXViewController


- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


#pragma mark MIXGameViewProtocol

- (BOOL)tryToDragPiecesFrom:(MIXCoreSquare)from
                         to:(MIXCoreSquare)to
                 withNumber:(NSUInteger)numberOfDraggedPieces {
    
    NSLog(@"Try to drag %lu pieces from %d/%d to %d/%d", (unsigned long)numberOfDraggedPieces,
          from.column, from.line, to.column, to.line);
    
    // if move not possible, [dragPieces...] does nothing
    BOOL movePossible = [self.board dragPiecesFrom:from to:to withNumber:numberOfDraggedPieces];
    [self.gameView clearBoard];
    [self.gameView setPiecesForBoard:self.board];
    
    return movePossible;
}

#pragma mark Methods

- (MIXModelBoard *)boardForTryingOut
{
    MIXModelBoard *board = [[MIXModelBoard alloc] init];
    for (int i = 0; i < 5; i++) {
        for (int j = 0; j < 5; j++) {
            [board setPiece:MIXCoreSquareMake(i, j)];
        }
    }
    
    for (int i = 1; i <=2; i++) {
        for (int j = 0; j <= 3; j++) {
            [board dragPiecesFrom:MIXCoreSquareMake(i, j)
                               to:MIXCoreSquareMake(i, j+1)
                       withNumber:j+1];
        }
    }
    return board;
}

@end

File: mixtour/MIXViewController.swift
//
//  MIXViewController.swift
//  mixtour
//
//  Created by Wenderdel, Carsten on 09/11/14.
//  Copyright (c) 2014 Carsten Wenderdel. All rights reserved.
//

import Foundation
import UIKit
import MixModel

class MIXViewController: UIViewController, GameViewDelegate {

    lazy var board: ModelBoard = ModelBoard()
    var gameView: GameView!
    var boardBeforeMove: ModelBoard?
    var undoButton: UIButton?

    
    override func viewDidLoad() {
        super.viewDidLoad()
        gameView = GameView(frame: self.view.frame)
        gameView.delegate = self
        gameView.setPiecesForBoard(self.board)
        self.view.addSubview(gameView)
        
        let theUndoButton = UIButton(type: .system)
        theUndoButton.frame = CGRect(x: 0, y: 30, width: 50, height: 15)
        theUndoButton.setTitle("Undo", for: UIControl.State())
        theUndoButton.addTarget(self, action: #selector(undoMove), for: .touchUpInside)
        theUndoButton.isEnabled = (nil != boardBeforeMove)
        self.view.addSubview(theUndoButton);
        self.undoButton = theUndoButton;
        
        let restartButton = UIButton(type: .system)
        restartButton.frame = CGRect(x: 70, y: 30, width: 50, height: 15)
        restartButton.setTitle("New", for: UIControl.State())
        restartButton.addTarget(self, action: #selector(restartGame), for: .touchUpInside)
        self.view.addSubview(restartButton);
    }
    
    
    @objc func undoMove() {
        if let undoBoard = boardBeforeMove {
            boardBeforeMove = nil;
            undoButton?.isEnabled = false
            board = undoBoard
        }
        gameView.setPiecesForBoard(board)
    }
    
    @objc func restartGame() {
        self.board = ModelBoard()
        self.gameView .setPiecesForBoard(self.board)
        self.undoButton?.isEnabled = false
    }
    
    private func calculateNextMoveForGameView(_ gameView: GameView) {
        
        DispatchQueue.global(qos: DispatchQoS.QoSClass.background).async(execute: { [weak self] in
            let move = self?.board.bestMove()
            
            if let bestMove = move {
                DispatchQueue.main.async(execute: {
                    // TODO: self.board could be changed by user during AI calculation - freeze parts of the UI or make a deep copy of the board
                    // Also the ownership of gameView and board is not thought through
                    guard let this = self else {return}
                    this.board.makeMoveIfLegal(bestMove)
                    gameView.setPiecesForBoard(this.board)
                })
            }
        })
    }
    

    // MARK: GameViewDelegate

    func gameView(_ gameView : GameView, tryToDragPieces numberOfDraggedPieces: Int, from: ModelSquare, to: ModelSquare) -> Bool {
        
        print("Try to drag \(numberOfDraggedPieces) pieces from \(from.column)/\(from.line) to \(to.column)/\(to.line)")
        
        let move = ModelMove(from: from, to: to, numberOfPieces: numberOfDraggedPieces)
        return self.gameView(gameView, tryToMakeMove: move)
    }
    
    func gameView(_ gameView : GameView, tryToSetPieceTo to: ModelSquare) -> Bool {
        
        print("Try to set piece to \(to.column)/\(to.line)")
        
        let move = ModelMove(setPieceTo: to)
        return self.gameView(gameView, tryToMakeMove: move)
    }
    
    func gameView(_ gameView : GameView, tryToMakeMove move: ModelMove) -> Bool {
        
        let oldBoard = ModelBoard(board: board)
        let movePossible = board.makeMoveIfLegal(move)
        
        // display new state. If move not possible, this also moves the dragged piece to the old correct position
        gameView.setPiecesForBoard(self.board)
        
        if (movePossible) {
            boardBeforeMove = oldBoard
            undoButton?.isEnabled = (nil != boardBeforeMove)
            calculateNextMoveForGameView(gameView)
        }
        
        return movePossible
    }

}


